database/tables.py

```diff
@@ -0,0 +1,141 @@
+        raise IndexError("%s is not within the week range [1, 52]" % week)
+
+    start = tz.localize(
+        datetime.strptime(f"{year}-W{week - 1}-1", "%Y-W%W-%w"))
+    return start, start + timedelta(days=7)
+
+
+class EventsTable:
+    """Interface to the event storage system.
+
+    All accessors on this table will dynamically generate the next recuring
+    events in the database based on the queried time range. For instance,
+
+    All dates are stored in UTC for ease of manipulation. This becomes relevant
+    whenever querying elements within a certain date range (e.g. weekly).
+    """
+
+    TABLE = "events"
+
+    # Record shape, for hint only.
+    RECORD = {
+        'date': datetime,
+        'title': str,
+        'description': str,
+        'repetition': bool,
+        'parent': Optional[datetime],
+    }
+
+    def __init__(self, db: Database):
+        """Constructor."""
+        self.db = db
+
+    def weekly(self, week: Optional[int] = None, year: Optional[int] = None,
+               tz: StandardTimezone = StandardTimezone.utc) -> List[Event]:
+        """Returns the list of event happening during the request week.
+
+        If no specific week is requested, this will return the events happening
+        during the current week.
+
+        :param week: The week number to query (1 to 52). Defaults to the
+                     current week.
+        :param year: The year to query. Defaults to the current year.
+        :param tz: Optional localization timezone to compute the current week
+                   and year.
+        """
+        now = datetime.now()
+        if not year:
+            year = now.year
+        if not week:
+            week = now.isocalendar()[1]
+
+        start, end = week_time_range(year, week, tz.value)
+        events = self._list_events(start, end, generate=True)
+        logging.debug("weekly: [%d, %d] -> [%s - %s]: %s",
+                      year, week, start.isoformat(), end.isoformat(), events)
+        return events
+
+    def save(self, event: Event):
+        """Saves an event in the table.
+
+        In the case the event is repeated, it will verify if the parent have
+        been stored in the database. If not, it will raise an exception.
+
+        :param event: Event to be stored.
+        """
+        if event.parent is not None:
+            if self.get(event.parent) is None:
+                raise AttributeError(
+                    "Event %s referes to an inexistent parent: %s" % (
+                        repr(event), event.parent))
+        self._save_record(event)
+
+    # Table handlers.
+
+    def get(self, date: datetime) -> Optional[Event]:
+        """Retrieves an event by its date. Returns None if no event is found.
+
+        :param date: an aware datetime object (with an associated timezone).
+        """
+        table = self.db.get_table(self.TABLE)
+        query = Query()
+        records = table.search(query.date == date)
+        if len(records) > 1:
```

Comment:
What is preventing the bot to input 2 events at the same time and therefore triggering this error everytime he tries to retrieve this event?

Reply:
Nice catch. I added a synchronized section ensuring this doesn't happen.